framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/RSPServerSocketLauncher.java

```diff
@@ -0,0 +1,127 @@
+			PrintWriter tracing) throws IOException {
+		super(localService, remoteInterface, socket, tracing);
+	}
+
+	
+	/**
+	 * Start a thread that listens for messages in the message producer and forwards them to the message consumer.
+	 * 
+	 * @param messageProducer - produces messages, e.g. by reading from an input channel
+	 * @param messageConsumer - processes messages and potentially forwards them to other consumers
+	 * @param executorService - the thread is started using this service
+	 * @return a future that is resolved when the started thread is terminated, e.g. by closing a stream
+	 */
+
+	@Override
+	protected Future<Void> startListeningCalled(StreamMessageProducer messageProducer, MessageConsumer messageConsumer,
+			ExecutorService execService, Object remoteProxy, RemoteEndpoint remoteEndpoint) {
+		
+		CustomMessageProcessor reader = new CustomMessageProcessor(messageProducer, messageConsumer, remoteProxy);
+		final Future<?> result = execService.submit(reader);
+		return toFutureVoid(result, messageProducer);
+	}
+	
+	private Future<Void> toFutureVoid(Future<?> result, StreamMessageProducer messageProducer) {
+		return new Future<Void>() {
+			@Override
+			public Void get() throws InterruptedException, ExecutionException {
+				return (Void) result.get();
+			}
+
+			@Override
+			public Void get(long timeout, TimeUnit unit)
+					throws InterruptedException, ExecutionException, TimeoutException {
+				return (Void) result.get(timeout, unit);
+			}
+
+			@Override
+			public boolean isDone() {
+				return result.isDone();
+			}
+
+			@Override
+			public boolean cancel(boolean mayInterruptIfRunning) {
+				if (mayInterruptIfRunning && messageProducer instanceof Closeable) {
+					try {
+						((Closeable) messageProducer).close();
+					} catch (IOException e) {
+						throw new RuntimeException(e);
+					}
+				}
+				return result.cancel(mayInterruptIfRunning);
+			}
+
+			@Override
+			public boolean isCancelled() {
+				return result.isCancelled();
+			}
+		};
+	}
+	
+	
+	public static class CustomMessageProcessor implements Runnable {
+		private static final Logger LOG = Logger.getLogger(ConcurrentMessageProcessor.class.getName());
+
+		private boolean isRunning;
+
+		private final MessageProducer messageProducer;
+		private final MessageConsumer messageConsumer;
+		private final Object remoteProxy;
+		
+		public CustomMessageProcessor(MessageProducer messageProducer, MessageConsumer messageConsumer, Object remoteProxy) {
+			this.messageProducer = messageProducer;
+			this.messageConsumer = messageConsumer;
+			this.remoteProxy = remoteProxy;
+		}
+
+		public void run() {
+			if (isRunning()) {
+				throw new IllegalStateException("The message processor is already running.");
+			}
+			setRunning(true);
+			try {
+				ClientThreadLocal.setActiveClient((RSPClient)remoteProxy);
+				messageProducer.listen(messageConsumer);
+			} catch (Exception e) {
```

Comment:
Is there more specific exception?

Reply:
No. I copied this code from upstream. It is important to catch every single exception here since these are run in Executor threads which may be re-used.  At least that's my understanding of the situation.